src/mpid/ch4/netmod/ucx/ucx_am.h

```diff
@@ -72,25 +69,16 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_NM_am_isend(int rank,
     ucx_hdr.handler_id = handler_id;
     ucx_hdr.data_sz = data_sz;
 
-    if (dt_contig) {
-        /* just pack and send for now */
-        send_buf = MPL_malloc(data_sz + am_hdr_sz + sizeof(ucx_hdr), MPL_MEM_BUFFER);
-        MPIR_Memcpy(send_buf, &ucx_hdr, sizeof(ucx_hdr));
-        MPIR_Memcpy(send_buf + sizeof(ucx_hdr), am_hdr, am_hdr_sz);
-        MPIR_Memcpy(send_buf + am_hdr_sz + sizeof(ucx_hdr), (char *) data + dt_true_lb, data_sz);
-    } else {
-        send_buf = MPL_malloc(data_sz + am_hdr_sz + sizeof(ucx_hdr), MPL_MEM_BUFFER);
-
-        MPIR_Memcpy(send_buf, &ucx_hdr, sizeof(ucx_hdr));
-        MPIR_Memcpy(send_buf + sizeof(ucx_hdr), am_hdr, am_hdr_sz);
+    send_buf = MPL_malloc(data_sz + am_hdr_sz + sizeof(ucx_hdr), MPL_MEM_BUFFER);
+    MPIR_Memcpy(send_buf, &ucx_hdr, sizeof(ucx_hdr));
+    MPIR_Memcpy(send_buf + sizeof(ucx_hdr), am_hdr, am_hdr_sz);
 
-        MPI_Aint actual_pack_bytes;
-        mpi_errno =
-            MPIR_Typerep_pack(data, count, datatype, 0, send_buf + am_hdr_sz + sizeof(ucx_hdr),
-                              data_sz, &actual_pack_bytes);
-        MPIR_ERR_CHECK(mpi_errno);
-        MPIR_Assert(actual_pack_bytes == data_sz);
-    }
+    MPI_Aint actual_pack_bytes;
+    mpi_errno =
+        MPIR_Typerep_pack(data, count, datatype, 0, send_buf + am_hdr_sz + sizeof(ucx_hdr),
```

Comment:
I guess one question is would `MPIR_Localcopy` be better? That is what we use in collectives when copying data to a temporary buffer. Internally, memcpy or typerep is used when needed.

Reply:
E.g. https://github.com/pmodels/mpich/blob/ccd708ee2375bd16b02e4a5e4154ca662021f270/src/mpi/coll/allgather/allgather_intra_brucks.c#L50-L51

Reply:
Might just be splitting hairs...

Reply:
`MPIR_Localcopy` seems to be used in cases in which we have both sender and receiver buffer information. From what I can see that can take a non-contig buffer and copy it into another non-contig buffer. It will definitely work even in this case but isn't that an overkill?

Reply:
That's true. In fact, seems like the usage I highlighted would make a good use-case for `MPIR_Typerep_pack` instead. My approval for the PR remains.

Reply:
We cannot use memcpy when it's a GPU buffer.  We should update `MPIR_Localcopy` to not do that.  Typerep is internally optimized for contiguous buffers anyway.

Reply:
@pavanbalaji I can update `MPIR_Localcopy` too.